nearcore/src/runtime/mod.rs

```diff
@@ -528,19 +534,45 @@ impl NightshadeRuntime {
         // Sort the receipts into appropriate outgoing shards.
         let mut receipt_result = HashMap::default();
         // Outgoing receipts should be sorted by shards of the epoch of the next block
-        let next_block_epoch_id = if self.is_next_block_epoch_start(prev_block_hash)? {
-            self.get_next_epoch_id_from_prev_block(prev_block_hash)?
-        } else {
-            self.get_epoch_id_from_prev_block(prev_block_hash)?
+        let next_shard_layout = {
+            let next_block_epoch_id = if self.is_next_block_epoch_start(prev_block_hash)? {
+                self.get_next_epoch_id_from_prev_block(prev_block_hash)?
+            } else {
+                self.get_epoch_id_from_prev_block(prev_block_hash)?
+            };
+            self.get_shard_layout(&next_block_epoch_id)?
         };
+        let shard_uid = self.get_shard_uid_from_prev_hash(shard_id, prev_block_hash)?;
+        let (consolidated_state_changes, split_state_apply_results) =
+            if self.will_shard_layout_change(prev_block_hash)? {
+                let consolidated_state_changes =
+                    ConsolidatedStateChanges::from_raw_state_changes(&apply_result.state_changes);
+                // split states are ready, apply update to them now
+                if let Some(state_roots) = split_state_roots {
+                    let split_state_results = Some(self.apply_update_to_split_states(
+                        block_hash,
+                        shard_uid.clone(),
+                        state_root,
+                        state_roots,
+                        &next_shard_layout,
+                        consolidated_state_changes,
+                    )?);
+                    (None, split_state_results)
+                } else {
+                    // split states are not ready yet, store state changes in consolidated_state_changes
+                    (Some(consolidated_state_changes), None)
+                }
+            } else {
+                (None, None)
+            };
```

Comment:
It might make sense to refactor this out into its own function

Reply:
I'm debating whether to move this part outside of runtime and to chain, I'll leave it as it is right now and do the refactoring at the end